onnxruntime/contrib_ops/cuda/collective/custom_reduce_impl.cu

```diff
@@ -0,0 +1,644 @@
+/*
+ * Copyright (c) 2022-2024, NVIDIA CORPORATION.  All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/providers/cuda/cu_inc/common.cuh"
+#include "core/providers/shared_library/provider_api.h"
+#include "custom_reduce_impl.h"
+#include <algorithm>
+#include <cstdint>
+#include <tuple>
+#include <type_traits>
+
+namespace ort_trtllm {
+
+#if defined(USE_MPI) || defined(USE_NCCL)
+
+using namespace onnxruntime;
+using namespace onnxruntime::cuda;
+
+// Calculates ceil(a / b). User must be careful to ensure that there
+// is no overflow or underflow in the calculation.
+template <typename T>
+constexpr T divUp(T a, T b) { return (a + b - (T)1) / b; }
+
+// Rounds a up to the next highest multiple of b. User must be careful
+// to ensure that there is no overflow or underflow in the calculation
+// of divUp.
+template <typename T>
+constexpr T roundUp(T a, T b) { return divUp<T>(a, b) * b; }
```

Comment:
Move divUp and roundUp to onnxruntime/core/providers/cuda/shared_inc/cuda_utils.h